src/box/lua/connpool.lua

```diff
@@ -125,6 +125,42 @@ local function connect_to_candidates(candidates)
     return connected_candidates
 end
 
+local acquire_any_cond = nil
+local function acquire_any_candidate(instance_names, opts)
+    assert(type(opts) == 'table' and opts.mode)
+    assert(opts.mode == 'ro' or opts.mode == 'rw' or opts.mode =='alive')
+    local time_connect_end = clock.monotonic() + WATCHER_TIMEOUT
+
+    if not acquire_any_cond then
+        acquire_any_cond = fiber.cond()
+    end
+
+    for _, instance_name in pairs(instance_names) do
+        local ok, conn = pcall(connect, instance_name, {
+            wait_connected = false,
+            connect_timeout = WATCHER_TIMEOUT
+        })
+
+        if ok then
+            conn:on_connect(function()
+                acquire_any_cond:broadcast()
+            end)
+        end
```

Comment:
`connect` may return a cached (already connected) connection. At least, we don't need the `on_connect` callback in the case.

Also, I'm wonder should we unregister the callback when it is unneeded? I guess it is saved and called on a reconnection. So, if we don't unregister it, it is present forever (if my guess it right). Maybe it is not a real problem, just noted to don't forget to think about it.

Reply:
Thank you.
I think you're right that the callback should be dropped. Added dropping the callback inside the callback.
Also, now the hook isn't set in case of the connection was already checked.

Reply:
> Also, now the hook isn't set in case of the connection was already checked.

OK.

> I think you're right that the callback should be dropped. Added dropping the callback inside the callback.

It seems, you tried to drop the trigger and hit #10622. OK to don't do it so.